From 667ae0b112cd636184433eb34ec16565c4ddb810 Mon Sep 17 00:00:00 2001
From: Jimmy Durand Wesolowski <jimmy.durand-wesolowski@openwide.fr>
Date: Tue, 2 Feb 2016 15:07:44 +0100
Subject: [PATCH 4/4] Correct submodule repository address.

---
 .gitmodules | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 5865ff9..bad7ddd 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,9 +1,9 @@
 [submodule "tools/git2cl"]
 	path = tools/git2cl
-	url = http://repo.or.cz/r/git2cl.git
+	url = git://repo.or.cz/git2cl.git
 [submodule "jimtcl"]
 	path = jimtcl
-	url = http://repo.or.cz/r/jimtcl.git
+	url = git://repo.or.cz/jimtcl.git
 [submodule "src/jtag/drivers/libjaylink"]
 	path = src/jtag/drivers/libjaylink
-	url = http://repo.or.cz/r/libjaylink.git
+	url = git://repo.or.cz/libjaylink.git
-- 
2.4.6

